### ISSUE TITLE
Add negated types as completion to union and intersection types

### ISSUE THREAD

**Issue by PetaSoft** [microsoft/TypeScript#39063]
## Search Terms

TypeScript 3.9 typesystem, union types, intersection types, negated types, complement types

## Suggestion

TypeScript already offers union types and intersections types to compose new types from existing ones. These are set operations, and in mathematics there also exists the complement set operation to express 
A \ B (elements of set A without elements of set B). It would be very helpful to have that expressiveness in TypeScript, too, and there are cases where there is no way to express a type other than by a complement type (or in other words: negated type).

Suggested syntax: `!Type` where `Type` is an already existing type.

## Use Cases

1) Sometimes it is necessary to express that the return value of a function/method is not of a conrete type. An Example is the static method `Object.create` (see below).
2) classes are constructable but not callable. It would be helpful to define an interface which exactly expresses these circumstances. Such an interface can be used in the `extends` clause of a type parameter.

## Examples

1) `Object.create` is currently declared by the overloaded method signatures:
```typescript
interface ObjectConstructor {
  create(o: object | null): any;
  create(o: object | null, properties: PropertyDescriptorMap & ThisType<any>): any;
}
```
The return type is `any`, which lacks type safety. One can improve the type safety by the following daclarations:
```typescript
interface ObjectConstructor {
  create<T extends object | null>(o: T): null extends T ? object: T;
  create<T extends object | null, U extends PropertyDescriptorMap & ThisType<any>>
    (o: T, properties: U): null extends T ? object : 
    { [P in keyof U]: U[P] extends { 'value': infer Q } ? Q : 
    U[P] extends { get(): infer R} ? R : 
    U[P] extends { set(v: infer S): void } ? S : unknown } & T;
}
```
This is already a much type safer version of `ObjectConstructor.create()` but in the case of `o` equal to `null` one get the flaw that the return type is `object`. Espacially, VS Code offers IntelliSence for values of type `object` which are the same as for type `Object` (big 'O' here) but the created object does not have any properties even not inherited properties. Futhermore, values `v` of type `object` can be assigned to variables `x` of type `Object`, but then `x instanceof Object` should be `true` which is not the case for `v`. One could switch from return type `object` to return type `unknown`. In that case the returned value is no longer assignable to a variable of type `Object` but that also does not reflect the fact that the returned value definitely is an object.
Therefore, it would be helpful to tell the compiler that the returned value is not of type `Object` but of type `object`, which implies that it cannot be assigned to variables of type `Object`. Currently, there is no way to prevent assignments of objects to `Object` variables but with negated types it is possible:
```typescript
interface ObjectConstructor {
  create<T extends object | null>(o: T): null extends T ? object & !Object : T;
  create<T extends object | null, U extends PropertyDescriptorMap & ThisType<any>>
    (o: T, properties: U): null extends T ? object & !Object : 
    { [P in keyof U]: U[P] extends { 'value': infer Q } ? Q : 
    U[P] extends { get(): infer R} ? R : 
    U[P] extends { set(v: infer S): void } ? S : unknown } & T;
}
```

2) classes are constructable but not callable. With negated types one can define an interface which exactly expresses these circumstances:
```typescript
interface classType {
  new (...args: any): any;
  ! (...args: any): any;
}
```
This `interface` can be used in `extends` clauses of type parameters:
```typescript
type ClassConstructorParameters<T extends classType> = 
        T extends new (...args: infer P) => any ? P : never;
```
## Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

**Comment by MartinJohns** [comment]
Duplicate of #29317.

**Comment by PetaSoft** [comment]
@MartinJohns  Thank You for the hint. I searched for the topic in the list open issues before I posted my feature request but did not find it because it is a pull request which is not part of the `is:issue is:open` list. Unfortunately, `is:issue is:open` is the default setting for search. Maybe that is the reason for some duplicates in the TypeScript issue tracker. In the future I will widen my search.

I will now close the issue.
